component/remote/s3/watcher.go

```diff
@@ -100,10 +99,15 @@ func (w *watcher) getObject(ctx context.Context) ([]byte, error) {
 	if err != nil {
 		return []byte{}, err
 	}
+	defer output.Body.Close()
+
 	buf := make([]byte, output.ContentLength)
-	_, err = output.Body.Read(buf)
-	if !errors.Is(err, io.EOF) {
```

Comment:
I remember what the problem with this is: the issue isn't an unexpected EOF, but rather a partial read.

Implementation of io.Reader are permitted to do partial reads; which is especially common over the network where files may be streaming. The documentation for io.Reader states:

> Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors. 

That wasn't happening here; we were assuming the first call to Read would always read the full bytes, which isn't required by the io.Reader implementation. Partial reads don't return errors, which means that on a partial read here we were hitting this condition (as `!errors.Is(nil, io.EOF)` is true). 

The new code guarantees that we'll fully consume the reader and completely fill our buffer to its length or return an error if we received EOF before filling up the buffer.